src/nma/nma.rs

```diff
@@ -0,0 +1,412 @@
+/// NAME Mips Assembler
+use crate::args::Args;
+use crate::u5::U5;
+use crate::u6::U6;
+use std::fs;
+use std::fs::File;
+use std::io::Write;
+use std::str;
+
+/// Number of expected arguments for R-type instructions
+const R_EXPECTED_ARGS: usize = 3;
+
+/// The form of an R-type instruction, specificially
+/// which arguments it expects in which order
+enum RForm {
+    None,
+    RdRsRt,
+    // Rs,
+    RdRtShamt,
+}
+
+/// The variable components of an R-type instruction
+pub struct R {
+    shamt: U5,
+    funct: U6,
+    form: RForm,
+}
+
+/// Number of expected arguments for I-type instructions
+const I_EXPECTED_ARGS: usize = 2;
+
+/// The form of an I-type instruction, specifically
+/// which arguments it expects in which order
+enum IForm {
+    None,
+    RtImm,
+}
+
+/// The variable components of an I-type instruction
+pub struct I {
+    opcode: U6,
+    form: IForm,
+}
+
+/// Parses an R-type instruction mnemonic into an [R]
+pub fn r_operation(mnemonic: &str) -> Result<R, &'static str> {
+    match mnemonic {
+        "add" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x20 },
+            form: RForm::RdRsRt,
+        }),
+        "sub" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x22 },
+            form: RForm::RdRsRt,
+        }),
+        "sll" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x00 },
+            form: RForm::RdRtShamt,
+        }),
+        "srl" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x02 },
+            form: RForm::RdRtShamt,
+        }),
+        "xor" => Ok(R {
+            shamt: U5 { value: 0 },
+            funct: U6 { value: 0x26 },
+            form: RForm::RdRsRt,
+        }),
+        _ => Err("Failed to match R-instr mnemonic"),
+    }
+}
+
+/// Parses an I-type instruction mnemonic into an [I]
+pub fn i_operation(mnemonic: &str) -> Result<I, &'static str> {
+    match mnemonic {
+        "lui" => Ok(I {
+            opcode: U6 { value: 0xf },
+            form: IForm::RtImm,
+        }),
+        _ => Err("Failed to match I-instr mnemonic"),
+    }
+}
+
+/// Split a string into meaningful, atomic elements of the MIPS language
+pub fn tokenize(raw_text: &str) -> Vec<&str> {
+    raw_text.split_whitespace().collect::<Vec<&str>>()
```

Comment:
Needs better matching of delimiters for greater ergonomics while writing code (this one is a big deal)

Reply:
Accounted for all whitespace and comma

Reply:
This does allow for statements like `add $t0 $t1 $t2` but I'm also going to resolve that with the real parser later